src/program.cpp

```diff
@@ -883,21 +891,42 @@ void Program::KernelFreed()
     --m_NumLiveKernels;
 }
 
-cl_int Program::ParseOptions(const char* optionsStr, CommonOptions& optionsStruct, bool SupportCompilerOptions, bool SupportLinkerOptions)
+void Program::AddBuiltinOptions(std::vector<Device::ref_ptr_int> const& devices, CommonOptions& optionsStruct)
 {
-    using namespace std::string_view_literals;
-
-    if (SupportCompilerOptions)
+    optionsStruct.Args.reserve(15);
+#ifdef CLON12_SUPPORT_3_0
+    optionsStruct.Args.push_back("-D__OPENCL_VERSION__=300");
+#else
+    optionsStruct.Args.push_back("-D__OPENCL_VERSION__=120");
+#endif
+    // Disable extensions promoted to optional core features that we don't support
+    optionsStruct.Args.push_back("-cl-ext=-cl_khr_fp64");
+    optionsStruct.Args.push_back("-cl-ext=-cl_khr_depth_images");
+    optionsStruct.Args.push_back("-cl-ext=-cl_khr_subgroups");
+    // Disable optional core features that we don't support
+    optionsStruct.Args.push_back("-cl-ext=-__opencl_c_fp64");
+    optionsStruct.Args.push_back("-cl-ext=-__opencl_c_device_enqueue");
+    optionsStruct.Args.push_back("-cl-ext=-__opencl_c_generic_address_space");
+    optionsStruct.Args.push_back("-cl-ext=-__opencl_c_pipes");
+    optionsStruct.Args.push_back("-cl-ext=-__opencl_c_program_scope_global_variables");
+    optionsStruct.Args.push_back("-cl-ext=-__opencl_c_subgroups");
+    optionsStruct.Args.push_back("-cl-ext=-__opencl_c_work_group_collective_functions");
+    // Query device caps to determine additional things to enable and/or disable
+    if (std::any_of(devices.begin(), devices.end(), [](Device::ref_ptr_int const& d) { return d->IsMCDM(); }))

```

Comment:
This means we lose image support even when runninmg on a real GPU if there's also an MCDM adapter in the system?

Reply:
This is saying that if you create a CL context which contains both a GPU and an MCDM device, and then issue a compile/build against that context targeting both the GPU and MCDM device, then that compile won't have image support. In order to have image support, you need to either:
* Only choose the GPU(s) when creating the context
* Explicitly specify the list of devices to compile/build for and exclude MCDM devices.